icicle/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,663 @@
+            bkt_idx = m_num_bkts * k + curr_coeff;
+          } else {
+            bkt_idx = m_num_bkts * k + ((-curr_coeff) & coeff_bit_mask_no_sign_bit);
+          }
+
+          // Check for collision in that bucket and either dispatch an addition or store the P accordingly.
+          if (m_bkts_occupancy[bkt_idx]) {
+            m_bkts_occupancy[bkt_idx] = false;
+            phase1_push_addition(bkt_idx, m_buckets[bkt_idx], carry > 0 ? A::neg(base) : base);
+          } else {
+            A base = m_are_points_mont ? A::from_montgomery(bases[m_precompute_factor * i + j])
+                                       : bases[m_precompute_factor * i + j];
+            if (negate_p_and_s) { base = A::neg(base); }
+            m_bkts_occupancy[bkt_idx] = true;
+            m_buckets[bkt_idx] = carry > 0 ? P::neg(P::from_affine(base)) : P::from_affine(base);
+          }
+        } else {
+          // Handle edge case where coeff = 1 << c due to carry overflow which means:
+          // coeff & coeff_mask == 0 but there is a carry to propagate to the next segment
+          carry = curr_coeff >> m_c;
+        }
+      }
+    }
+  }
+  phase1_wait_for_completion();
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase1_push_addition(const unsigned int task_bkt_idx, const P bkt, const A base)
+{
+  EcAddTask<A, P>* task = nullptr;
+  while (task == nullptr) {
+    // Use the search for an available (idle or completed) task as an opportunity to handle the existing results.
+    task = manager.get_idle_or_completed_task();
+    if (task->is_completed()) {
+      // Check for collision in the destination bucket, and chain and addition / store result accordingly.
+      if (m_bkts_occupancy[task->m_return_idx]) {
+        m_bkts_occupancy[task->m_return_idx] = false;
+        task->set_phase1_collision_task(m_buckets[task->m_return_idx]);
+        task = nullptr;
+        continue;
+      } else {
+        m_buckets[task->m_return_idx] = task->m_p1;
+        m_bkts_occupancy[task->m_return_idx] = true;
+      }
+    }
+    // After handling the result a new one can be set.
+    task->set_phase1_addition_with_affine(bkt, base, task_bkt_idx);
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase1_wait_for_completion()
+{
+  EcAddTask<A, P>* task = manager.get_completed_task();
+  while (task != nullptr) {
+    // Check for collision in the destination bucket, and chain and addition / store result accordingly.
+    if (m_bkts_occupancy[task->m_return_idx]) {
+      m_bkts_occupancy[task->m_return_idx] = false;
+      task->set_phase1_collision_task(m_buckets[task->m_return_idx]);
+    } else {
+      m_buckets[task->m_return_idx] = task->m_p1;
+      m_bkts_occupancy[task->m_return_idx] = true;
+      task->set_idle();
+    }
+    task = manager.get_completed_task();
+  }
+}
+
+template <typename A, typename P>
+void Msm<A, P>::phase2_bm_sum(std::vector<BmSumSegment>& segments)
+{
+  phase2_setup(segments);
```

Comment:
as we discussed this is probably extraneous

Reply:
will be dealt within  a future version